src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs

```diff
@@ -0,0 +1,399 @@
+    /// retrying.
+    pub connect_timeout: Duration,
+
+    /// The TCP user timeout for a Postgres/CRDB connection. Specifies the
+    /// amount of time that transmitted data may remain unacknowledged before
+    /// the TCP connection is forcibly closed.
+    pub tcp_user_timout: Duration,
+}
+
+impl From<PostgresTimestampOracleConfig> for PostgresClientConfig {
+    fn from(config: PostgresTimestampOracleConfig) -> Self {
+        let metrics = config.metrics.postgres_client.clone();
+        PostgresClientConfig::new(config.url.clone(), Arc::new(config), metrics)
+    }
+}
+
+impl PostgresTimestampOracleConfig {
+    /// Returns a new instance of [`PostgresTimestampOracleConfig`] with default tuning.
+    pub fn new(url: &str, metrics: Arc<Metrics>) -> Self {
+        PostgresTimestampOracleConfig {
+            url: url.to_string(),
+            metrics,
+            // NOTE: These defaults are taken as is from the persist Consensus
+            // tuning. Once we're sufficiently advanced we might want to a) make
+            // some of these configurable dynamically, as they are in persist,
+            // and b) pick defaults that are different from the persist
+            // defaults.
+            connection_pool_max_size: 50,
+            connection_pool_ttl: Duration::from_secs(300),
+            connection_pool_ttl_stagger: Duration::from_secs(6),
+            connect_timeout: Duration::from_secs(5),
+            tcp_user_timout: Duration::from_secs(30),
+        }
+    }
+}
+
+impl PostgresClientKnobs for PostgresTimestampOracleConfig {
+    fn connection_pool_max_size(&self) -> usize {
+        self.connection_pool_max_size
+    }
+
+    fn connection_pool_ttl(&self) -> Duration {
+        self.connection_pool_ttl.clone()
+    }
+
+    fn connection_pool_ttl_stagger(&self) -> Duration {
+        self.connection_pool_ttl_stagger.clone()
+    }
+
+    fn connect_timeout(&self) -> Duration {
+        self.connect_timeout.clone()
+    }
+
+    fn tcp_user_timeout(&self) -> Duration {
+        self.tcp_user_timout.clone()
+    }
+}
+
+impl<N> PostgresTimestampOracle<N>
+where
+    N: GenericNowFn<Timestamp> + 'static,
+{
+    /// Open a Postgres [`TimestampOracle`] instance with `config`, for the
+    /// timeline named `timeline`. `next` generates new timestamps when invoked.
+    /// Timestamps that are returned are made durable and will never retract.
+    pub(crate) async fn open(
+        config: PostgresTimestampOracleConfig,
+        timeline: String,
+        initially: Timestamp,
+        next: N,
+    ) -> Result<Self, anyhow::Error> {
+        info!(url = ?config.url, "opening PostgresTimestampOracle");
+
+        let metrics = Arc::clone(&config.metrics);
+
+        let postgres_client = PostgresClient::open(config.into())?;
+
+        let client = postgres_client.get_connection().await?;
+
+        // WIP: Do we need this for the pg oracle? This is originally from
```

Comment:
Maybe! The comment isn't quite true of this impl (this modifies a small set of rows instead of creating and deleting new rows) but I think GC would still be an issue. The main disadvantage to setting this lower is that we can't use incremental backups, but not sure that gets us a ton here anyway. (A secondary disadvantage is that it's a pretty atypical setting to use and will likely get pointed out by CRL support and have to be explained if/when we escalate anything, but we're already getting this one because persist.)

Reply:
I'll leave it in then and remove the WIP comment